// @builder.io/sdk-vue/vue2/src/functions/register-component.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  components: () => components,
  registerComponent: () => registerComponent
});
var import_is_browser = __toModule(require("./is-browser"));
const components = {};
function registerComponent(component, info) {
  var _a;
  components[info.name] = { component, info };
  if ((0, import_is_browser.isBrowser)()) {
    const sendInfo = prepareComponentInfoToSend(info);
    (_a = window.parent) == null ? void 0 : _a.postMessage({
      type: "builder.registerComponent",
      data: sendInfo
    }, "*");
  }
  return component;
}
function prepareComponentInfoToSend(info) {
  return __spreadValues(__spreadValues({}, info), info.inputs && {
    inputs: info.inputs.map((input) => {
      const keysToConvertFnToString = ["onChange", "showIf"];
      for (const key of keysToConvertFnToString) {
        if (input[key] && typeof input[key] === "function") {
          const fn = input[key];
          input = __spreadProps(__spreadValues({}, input), {
            [key]: `return (${fn.toString()}).apply(this, arguments)`
          });
        }
      }
      return input;
    })
  });
}
